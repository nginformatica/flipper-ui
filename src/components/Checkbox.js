import React from 'react'
import _ from 'prop-types'
import styled from 'styled-components'
import { primary, background } from '../colors'

const Label = styled.label`
    padding: 12px;
    transition: all 500ms ease;
    font-family: 'Roboto', sans-serif;
    cursor: pointer;
    margin: 0.75em;
    border-radius: 6px;
    &:hover {
        color: ${primary.normal};
        background: ${background.light}
    }
`

const Input = styled.input`
    margin-right: ${props => props.label ? '0px' : '12px' };
`

const Checkbox = ({ label, style, onChange }) =>
    <Label style={ style }>
        <Input
            type="checkbox"
            onChange={ onChange }
        />
        { label }
    </Label>

Checkbox.propTypes = {
    label: _.string,
    style: _.object,
    onChange: _.func
}

export default Checkbox
