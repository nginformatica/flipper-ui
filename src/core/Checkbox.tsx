import {
    Checkbox as MuiCheckbox,
    FormControlLabel as MuiFormControlLabel,
    Switch as MuiSwitch,
    Typography
} from '@material-ui/core'
import React, { ChangeEvent, ReactNode } from 'react'
import { IDefault } from './Advertise'
import { HelperBox, TextFieldWrapper as CheckFieldsWrapper } from './TextField'
interface IProps extends IDefault {
    name: string
    label?: ReactNode
    color?: 'primary' | 'secondary' | 'default'
    className?: string
    disabled?: boolean
    checked?: boolean
    dense?: boolean
    type?: 'switch' | 'checkbox'
    helperIcon?: React.ReactNode
    onHelperClick?: () => void
    onChange?: (event: ChangeEvent<HTMLElement>) => void
}

const DENSE = { padding: '2px', margin: '0px 7px' }

const Checkbox = (props: IProps) => {
    const {
        type = 'checkbox',
        margin,
        padding,
        style,
        helperIcon,
        onHelperClick
    } = props

    const renderHelper = (
        <>
            {
                onHelperClick && (
                    <HelperBox
                        helperIcon={ helperIcon }
                        onHelperClick={ onHelperClick }
                    />
                )
            }
        </>
    )

    const renderCheckbox = () =>
        <MuiCheckbox
            id={ props.id }
            name={ props.name }
            checked={ props.checked }
            value={ props.name }
            color={ props.color }
            style={ props.dense ? DENSE : {} }
            disabled={ props.disabled }
            onChange={ props.onChange }
        />

    const renderSwitch = () =>
        <MuiSwitch
            id={ props.id }
            name={ props.name }
            checked={ props.checked }
            value={ props.name }
            color={ props.color }
            disabled={ props.disabled }
            onChange={ props.onChange }
        />

    const renderControl = () => type === 'checkbox'
        ? renderCheckbox()
        : renderSwitch()

    const renderLabel = () =>
        typeof props.label === 'string' && props.dense
            ? (
                <Typography variant='body2'>
                    { props.label }
                </Typography>
            )
            : props.label

    return (
        <CheckFieldsWrapper>
            {
                props.label
                    ? (
                        <MuiFormControlLabel
                            style={ { padding, margin, ...style } }
                            className={ props.className }
                            label={ renderLabel() }
                            control={ renderControl() }
                        />
                    )
                    : renderControl()
            }
            { renderHelper }
        </CheckFieldsWrapper>
    )
}

export default Checkbox
