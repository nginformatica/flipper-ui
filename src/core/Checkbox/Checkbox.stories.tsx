import React from 'react'
import { Meta, StoryFn } from '@storybook/react'
import Checkbox from '.'

export default {
    title: 'Checkbox',
    component: Checkbox,
    argTypes: {
        backgroundColor: { control: 'color' }
    }
} as Meta<typeof Checkbox>

const Template: StoryFn<typeof Checkbox> = args => <Checkbox {...args} />

export const Default = () => (
    <Checkbox
        label='I agree with the terms'
        name='terms'
        onChange={() => alert('You clicked on the checkbox!')}
    />
)

export const WithHelper = Template.bind({})
WithHelper.args = {
    label: 'I agree with the terms',
    name: 'terms',
    onChange: () => alert('You clicked on the checkbox!'),
    onHelperClick: () => alert('Do you need help?'),
    children: 'Checkbox'
}

export const Primary = Template.bind({})
Primary.args = {
    label: 'I agree with the terms',
    name: 'terms',
    onChange: () => alert('You clicked on the checkbox!'),
    color: 'primary',
    children: 'Checkbox'
}

export const WithoutLabel = Template.bind({})
WithoutLabel.args = {
    name: 'terms',
    onChange: () => alert('You clicked on the checkbox!'),
    children: 'Checkbox'
}

export const Switch = Template.bind({})
Switch.args = {
    label: 'Bluetooth',
    name: 'bluetooth',
    type: 'switch',
    onChange: () => alert('You clicked on the switcher!'),
    children: 'Checkbox'
}

export const SwitchWithHelper = Template.bind({})
SwitchWithHelper.args = {
    id: 'with-helper',
    label: 'Bluetooth',
    name: 'bluetooth',
    type: 'switch',
    onChange: () => alert('You clicked on the switcher!'),
    onHelperClick: () => alert('Do you need help?'),
    children: 'Checkbox'
}
