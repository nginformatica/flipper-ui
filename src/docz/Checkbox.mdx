---
name: Checkbox
route: /checkbox
---

import Checkbox from '../components/Checkbox'
import { Playground, PropsTable } from 'docz'

# Checkbox

## Properties
<PropsTable of={ Checkbox } />

## Normal
<Playground>
    <Checkbox
        label="I agree with the terms"
        name="terms"
        onChange={ () => alert('You clicked on the checkbox!') }
    />
</Playground>


## Primary
<Playground>
    <Checkbox
        label="I agree with the terms"
        name="terms"
        color="primary"
        onChange={ () => alert('You clicked on the checkbox!') }
    />
</Playground>

## Without label
<Playground>
    <Checkbox
        name="terms"
        onChange={ () => alert('You clicked on the checkbox!') }
    />
</Playground>

## Switch
<Playground>
    <Checkbox
        name="bluetooth"
        type="switch"
        label="Bluetooth"
        onChange={ () => alert('You clicked on the switcher!') }
    />
</Playground>
