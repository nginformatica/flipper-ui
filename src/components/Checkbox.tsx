import { Checkbox as MuiCheckbox, FormControlLabel, Switch } from '@material-ui/core'
import React, { Component } from 'react'

interface IProps {
    name: string
    label?: string
    style?: object
    color?: 'primary' | 'secondary' | 'default'
    disabled?: boolean
    checked?: boolean
    type?: 'switch' | 'checkbox'
    onChange?: (event) => void
}

class Checkbox extends Component<IProps> {
    public static defaultProps = {
        type: 'checkbox'
    }

    public renderCheckbox() {
        return (
            <MuiCheckbox
                checked={ this.props.checked }
                value={ this.props.name }
                color={ this.props.color }
                disabled={ this.props.disabled }
                onChange={ this.props.onChange }
            />
        )
    }

    public renderSwitch() {
        return (
            <Switch
                checked={ this.props.checked }
                value={ this.props.name }
                color={ this.props.color }
                disabled={ this.props.disabled }
                onChange={ this.props.onChange }
            />
        )
    }

    public renderControl() {
        return this.props.type === 'checkbox'
            ? this.renderCheckbox()
            : this.renderSwitch()
    }

    public render() {
        const { label, style } = this.props

        return label
            ? (
                <FormControlLabel
                    style={ style }
                    label={ label }
                    control={ this.renderControl() }
                />
            )
            : this.renderControl()
    }
}

export default Checkbox
