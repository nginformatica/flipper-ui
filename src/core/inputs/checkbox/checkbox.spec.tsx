import React from 'react'
import { render, screen, act, waitFor, fireEvent } from '@testing-library/react'
import Checkbox from '.'

describe('Checkbox', () => {
    it('should render', () => {
        render(<Checkbox name='checkbox-name' onChange={jest.fn()} />)

        const checkbox = screen.getByRole('checkbox')

        expect(checkbox).toBeDefined()
    })

    it('should render with label', () => {
        render(
            <Checkbox
                label='checkbox-label'
                name='checkbox-name'
                onChange={jest.fn()}
            />
        )

        const label = screen.getByText('checkbox-label')

        expect(label).toBeDefined()
    })

    it('should render dense', () => {
        render(
            <Checkbox
                dense
                label='checkbox-label'
                name='checkbox-name'
                onChange={jest.fn()}
            />
        )

        const label = screen.getByText('checkbox-label')

        expect(label.classList).toContain('MuiTypography-body2')
    })

    it('should render switch', () => {
        render(
            <Checkbox
                type='switch'
                label='checkbox-label'
                name='checkbox-name'
                onChange={jest.fn()}
            />
        )

        const checkbox = screen.getByRole('switch')

        expect(checkbox).toBeDefined()
    })

    it('should render with helper', () => {
        render(
            <Checkbox
                label='checkbox-label'
                name='checkbox-name'
                onChange={jest.fn()}
                onHelperClick={jest.fn()}
            />
        )

        const helper = screen.getByRole('helper-box')

        expect(helper).toBeDefined()
    })

    it('should call onChange using checkbox', () => {
        const onChange = jest.fn()

        render(
            <Checkbox
                label='checkbox-label'
                name='checkbox-name'
                onChange={onChange}
            />
        )

        const checkbox = screen.getByRole('checkbox')

        act(() => {
            checkbox.click()
        })

        expect(onChange).toHaveBeenCalled()
    })

    it('should call onChange using switch', () => {
        const onChange = jest.fn()

        render(
            <Checkbox
                type='switch'
                label='checkbox-label'
                name='checkbox-name'
                onChange={onChange}
            />
        )

        const checkbox = screen.getByRole('switch')

        act(() => {
            checkbox.click()
        })

        expect(onChange).toHaveBeenCalled()
    })

    it('should call onHelperClick', async () => {
        const onHelperClick = jest.fn()

        render(
            <Checkbox
                label='checkbox-label'
                name='checkbox-name'
                onChange={jest.fn()}
                onHelperClick={onHelperClick}
            />
        )

        const helper = screen.getByRole('helper-box')
        const button = helper.querySelector('svg')

        fireEvent.click(button || helper)

        await waitFor(() => {
            expect(onHelperClick).toHaveBeenCalled()
        })
    })

    it('should match snapshot', () => {
        const { container } = render(
            <Checkbox
                type='switch'
                label='checkbox-label'
                name='checkbox-name'
                onChange={jest.fn()}
            />
        )

        expect(container).toMatchSnapshot()
    })
})
