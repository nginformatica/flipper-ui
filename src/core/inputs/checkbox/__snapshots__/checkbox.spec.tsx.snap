// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should match snapshot 1`] = `
<div>
  <div
    class="sc-iGgVNO csexcM"
  >
    <label
      class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
    >
      <span
        class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
      >
        <span
          class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary css-byenzh-MuiButtonBase-root-MuiSwitch-switchBase"
          role="switch"
        >
          <input
            class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
            name="checkbox-name"
            type="checkbox"
            value="checkbox-name"
          />
          <span
            class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
          />
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
        />
      </span>
      <span
        class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
      >
        checkbox-label
      </span>
    </label>
  </div>
</div>
`;
